File: solutions/209.minimum-size-subarray-sum/minimum-size-subarray-sum.java
class Solution {
    public int minSubArrayLen(int s, int[] nums) {
        int len = nums.length;
        int res = 0;
        int[] sums = new int[len + 1];
        sums[0] = 0;
        for(int i = 1; i <= len; i++){
            sums[i] = sums[i - 1] + nums[i - 1];
        }
        for(int i = 0; i < len; i++){
            int right = helper(sums, i + 1, len, sums[i] + s);
            if(right > len) break;
            res = res == 0 ?right - i : right - i > res ? res : right -i; 
        }
        return res;
    }
    
    public int helper(int[] sums,  int left, int right, int target){
        while(left <= right){
            int mid = left + (right - left) / 2;
            if(sums[mid] >= target) {
               right = mid - 1; 
            }else{
                left = mid + 1;
            }
        }
        return left;
    }
}

File: solutions/209.minimum-size-subarray-sum/minimum-size-subarray-sum.js
/**
 * @param {number} s
 * @param {number[]} nums
 * @return {number}
 */
var minSubArrayLen = function(s, nums) {
    if(nums.length === 0) return 0;
    let left = 0, right = 0, sum = 0;
    let res = Number.MAX_SAFE_INTEGER;
    while(right < nums.length){
        sum += nums[right];    
        while(sum >= s){
            res = Math.min(res, right - left + 1);
            sum -= nums[left];
            left++;
        } 
        right++;
    }
    if(res === Number.MAX_SAFE_INTEGER) return 0;
    return res;
};